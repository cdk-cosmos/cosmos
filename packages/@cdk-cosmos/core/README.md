# Cosmos Core Package

This Library contains the base components for the CDK Cosmos. It defiles the Core Classes and the Extension Classes for the Core.
All other @cdk-cosmos packages will build on top of this core concept.

## Basic Understanding

Cosmos is broken down into 3 scopes (Bubbles)

- Cosmos: Top level scope of the entire infrastructure, mostly singleton shared resources.
- Galaxy: Account level scope that defines infrastructure for an aws account.
- SolarSystem: Environment level scope that defines infrastructure for an environment with in an aws account (Dev, Tst, Prd).

## Cosmos Core

The Cosmos core contains infrastructure that will be shared by consumers building apps (extensions).

The Core Cosmos is composed of:

- CosmosStack
  - TLD Zone
  - CDK Repo - for the core cdk code
  - CDK Master Role - for builds of the cdk deployments
- GalaxyStack
  - If Cross Account, create Cross Account CDK Role for master to assume
- SolarSystemStack
  - VPC (Created on the Galaxy scope, so that it can be shared by many SolarSystems in the same Galaxy)
  - Zone
- EcsSolarSystemStack
  - Cluster
  - Alb
  - HttpListener
- CiCdSolarSystemStack
  - VPC
  - Zone
  - Cluster
  - Alb
  - HttpListener
  - CDK Pipeline + Code Build for Deployment of the core

### Example

```js
const cosmos = new CosmosStack(app, 'Demo', {
  tld: 'cosmos.com',
  cidr: '10.0.0.0/22',
});

const mgtGalaxy = new GalaxyStack(cosmos, 'Mgt');

const ciCd = new CiCdSolarSystemStack(mgtGalaxy);

const dev = new EcsSolarSystemStack(mgtGalaxy, 'Dev');
```

## Cosmos Extensions

The Extensions are a way of consuming and adding resources in an isolated way. Each Extension has a Portal Property that is used to consume that core infrastructure for that same bubble.

The Extension of the core are:

- CosmosExtensionStack
  - Imports Core Cosmos
  - CDK Repo - for the app cdk code
- GalaxyExtensionStack
  - Imports Core Galaxy
- SolarSystemExtensionStack
  - Imports Core SolarSystem
- EcsSolarSystemExtensionStack
  - Imports Core EcsSolarSystem
- CiCdSolarSystemExtensionStack
  - Imports Core CiCdSolarSystem
  - CDK Pipeline + Code Build for Deployment of the app

### Example

```js
const cosmos = new CosmosExtensionStack(app, 'Demo');

const mgtGalaxy = new GalaxyExtensionStack(cosmos, 'Mgt');

const ciCd = new CiCdSolarSystemExtensionStack(mgtGalaxy);

const dev = new EcsSolarSystemExtensionStack(mgtGalaxy, 'Dev');
```

What is included in the above is just enough for the extension (App) to host its own code and deploy it self. From here we expect consumer to add their own infrastructure to extends from the core, into the respective bubbles.

## Further Use

For more extensive understanding and usage, please look at the docs, aka the [law of the cosmos](https://github.com/cdk-cosmos/law)

Deploy CDKToolKit `npx cdk deploy --app="node_modules/@cosmos-building-blocks/common/lib/cdk-tool-kit-app.js"`