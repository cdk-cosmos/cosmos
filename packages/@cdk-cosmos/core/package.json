{
  "name": "@cdk-cosmos/core",
  "version": "0.2.0",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib/*",
    "package.json",
    "README.md"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cdk-cosmos/cosmos.git",
    "directory": "packages/@cdk-cosmos/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "test": "jest",
    "lint": "eslint . --ext .js,.ts",
    "format": "prettier --write **/*.ts !**/*.d.ts",
    "clean": "rm -rf lib"
  },
  "dependencies": {
    "@cosmos-building-blocks/common": "^0.2.0"
  },
  "peerDependencies": {
    "@aws-cdk/aws-cloudformation": "^1.23.0",
    "@aws-cdk/aws-codebuild": "^1.23.0",
    "@aws-cdk/aws-codecommit": "^1.23.0",
    "@aws-cdk/aws-codepipeline": "^1.23.0",
    "@aws-cdk/aws-codepipeline-actions": "^1.23.0",
    "@aws-cdk/aws-ec2": "^1.23.0",
    "@aws-cdk/aws-ecr": "^1.23.0",
    "@aws-cdk/aws-ecs": "^1.23.0",
    "@aws-cdk/aws-elasticloadbalancingv2": "^1.23.0",
    "@aws-cdk/aws-elasticloadbalancingv2-targets": "^1.23.0",
    "@aws-cdk/aws-iam": "^1.23.0",
    "@aws-cdk/aws-lambda": "^1.23.0",
    "@aws-cdk/aws-route53": "^1.23.0",
    "@aws-cdk/core": "^1.23.0"
  },
  "gitHead": "173c051038a3acdaa93d6a37d42b6b0546ffd567"
}
