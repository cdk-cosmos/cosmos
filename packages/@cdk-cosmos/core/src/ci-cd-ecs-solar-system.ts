import { Project } from '@aws-cdk/aws-codebuild';
import {
  IEcsSolarSystemCore,
  EcsSolarSystemCoreStack,
  EcsSolarSystemCoreProps,
  ImportedEcsSolarSystemCore,
  EcsSolarSystemExtensionStack,
  IEcsSolarSystemExtension,
} from './ecs-solar-system';
import { IGalaxyExtension, IGalaxyCore } from './galaxy';
import { CdkPipeline } from './components/cdk-pipeline';
import {
  ICiCdSolarSystemCore,
  ICiCdSolarSystemExtension,
  CiCdSolarSystemCoreStackProps,
  CiCdSolarSystemExtensionStackProps,
  CosmosCdkPipeline,
} from './ci-cd-solar-system';

export interface ICiCdEcsSolarSystemCore extends ICiCdSolarSystemCore, IEcsSolarSystemCore {}

export interface CiCdEcsSolarSystemCoreStackProps extends CiCdSolarSystemCoreStackProps, EcsSolarSystemCoreProps {}

export class CiCdEcsSolarSystemCoreStack extends EcsSolarSystemCoreStack implements ICiCdEcsSolarSystemCore {
  readonly deployPipeline: CdkPipeline;
  readonly deployProject: Project;

  constructor(galaxy: IGalaxyCore, props?: CiCdEcsSolarSystemCoreStackProps) {
    super(galaxy, 'CiCd', {
      description: 'Cosmos CiCdEcsSolarSystem: Resources dependant on Ci & Cd, like CodePipelines and CodeDeployments.',
      ...props,
    });

    const { cdkPipelineProps } = props || {};

    this.deployPipeline = new CosmosCdkPipeline(this, 'CdkPipeline', cdkPipelineProps);
    this.deployProject = this.deployPipeline.Deploy;
  }
}

export interface ICiCdEcsSolarSystemExtension extends ICiCdSolarSystemExtension, IEcsSolarSystemExtension {
  portal: ICiCdEcsSolarSystemCore;
}

export interface CiCdEcsSolarSystemExtensionStackProps extends CiCdSolarSystemExtensionStackProps {}

export class CiCdEcsSolarSystemExtensionStack extends EcsSolarSystemExtensionStack
  implements ICiCdEcsSolarSystemExtension {
  readonly portal: ICiCdEcsSolarSystemCore;
  readonly deployPipeline: CdkPipeline;
  readonly deployProject: Project;

  constructor(galaxy: IGalaxyExtension, props?: CiCdEcsSolarSystemExtensionStackProps) {
    super(galaxy, 'CiCd', {
      description:
        'Cosmos CiCdEcsSolarSystem Extension: App resources dependant on Ci & Cd, like CodePipelines and CodeDeployments.',
      ...props,
    });

    const { cdkPipelineProps } = props || {};

    this.deployPipeline = new CosmosCdkPipeline(this, 'CdkPipeline', cdkPipelineProps);
    this.deployProject = this.deployPipeline.Deploy;
  }

  protected getPortal(props?: CiCdSolarSystemExtensionStackProps): ICiCdEcsSolarSystemCore {
    const galaxy = this.node.scope as IGalaxyExtension;
    return new ImportedEcsSolarSystemCore(galaxy.portal, this.node.id, props?.portalProps);
  }
}
