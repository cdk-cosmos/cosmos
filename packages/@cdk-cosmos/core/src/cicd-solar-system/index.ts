export {
  ICiCdSolarSystemCore,
  CiCdSolarSystemCoreStack,
  CiCdEcsSolarSystemCoreStack,
  CiCdSolarSystemCoreStackProps,
} from './cicd-solar-system-core-stack';
export {
  ICiCdSolarSystemExtension,
  CiCdSolarSystemExtensionStack,
  CiCdEcsSolarSystemExtensionStack,
  CiCdSolarSystemExtensionStackProps,
} from './cicd-solar-system-extension-stack';
