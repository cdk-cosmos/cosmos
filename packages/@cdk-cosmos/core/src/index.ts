export {
  Bubble,
  Cosmos,
  Galaxy,
  SolarSystem,
  EcsSolarSystem,
  CiCdSolarSystem,
  Extension,
  CosmosExtension,
  GalaxyExtension,
  SolarSystemExtension,
  EcsSolarSystemExtension,
  CiCdSolarSystemExtension,
} from './interfaces';

export {
  RemoteZone,
  RemoteVpc,
  RemoteCluster,
  RemoteAlb,
  RemoteApplicationListener,
  RemoteCodeRepo,
  RemoteBuildProject,
} from './remote';

export { CosmosStack, CosmosStackProps, ImportedCosmos, CosmosExtensionStack } from './cosmos';

export { GalaxyStack, GalaxyStackProps, ImportedGalaxy, GalaxyExtensionStack } from './galaxy';

export { SolarSystemStack, SolarSystemProps, ImportedSolarSystem, SolarSystemExtensionStack } from './solar-system';

export {
  EcsSolarSystemStack,
  EcsSolarSystemProps,
  ImportedEcsSolarSystem,
  EcsSolarSystemExtensionStack,
} from './ecs-solar-system';

export {
  CiCdSolarSystemStack,
  CiCdStackProps,
  ImportedCiCdSolarSystem,
  CiCdSolarSystemExtensionStack,
} from './ci-cd-solar-system';

export { CdkPipeline, CdkPipelineProps } from './cdk-pipeline';

export { RESOLVE, PATTERN, _RESOLVE, ResolveParams } from './pattern';
