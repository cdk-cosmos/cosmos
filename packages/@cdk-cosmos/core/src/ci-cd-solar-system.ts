import { Construct, StackProps } from '@aws-cdk/core';

import { Project } from '@aws-cdk/aws-codebuild';
import { Role } from '@aws-cdk/aws-iam';
import {
  PATTERN,
  Galaxy,
  EcsSolarSystem,
  CiCdSolarSystem,
  GalaxyExtension,
  EcsSolarSystemStack,
  EcsSolarSystemProps,
  ImportedEcsSolarSystem,
  EcsSolarSystemExtensionStack,
  CiCdSolarSystemExtension,
  CdkPipeline,
  CdkPipelineProps,
} from '.';

export interface CiCdStackProps extends EcsSolarSystemProps {
  cdkPipelineProps?: Partial<CdkPipelineProps>;
}

export class CiCdSolarSystemStack extends EcsSolarSystemStack implements CiCdSolarSystem {
  readonly CdkDeploy: Project;

  constructor(galaxy: Galaxy, props?: CiCdStackProps) {
    super(galaxy, 'CiCd', props);

    const { cdkPipelineProps = {} } = props || {};
    const { CdkRepo, CdkMasterRoleStaticArn } = this.Galaxy.Cosmos;

    const pipeline = new CdkPipeline(this, 'CdkPipeline', {
      name: this.RESOLVE(PATTERN.SINGLETON_COSMOS, 'Cdk-Pipeline'),
      cdkRepo: CdkRepo,
      deployRole: Role.fromRoleArn(this, 'CdkMasterRole', CdkMasterRoleStaticArn, { mutable: false }),
      // deployVpc: this.Vpc,
      deployStacks: [this.RESOLVE(PATTERN.COSMOS, '*')],
      ...cdkPipelineProps,
    });
    this.CdkDeploy = pipeline.Deploy;

    // RemoteBuildProject.export(`Core${this.Account.Name}${this.Name}`, this.CdkDeploy);
  }
}

export class ImportedCiCdSolarSystem extends ImportedEcsSolarSystem implements CiCdSolarSystem {
  // readonly CdkDeploy: IProject;

  constructor(scope: Construct, galaxy: Galaxy) {
    super(scope, galaxy, 'CiCd');

    // this.CdkDeploy = RemoteBuildProject.import(this, `Core${this.Galaxy.Name}${this.Name}`, 'CdkPipelineDeploy');
  }
}

export class CiCdSolarSystemExtensionStack extends EcsSolarSystemExtensionStack implements CiCdSolarSystemExtension {
  readonly Portal: EcsSolarSystem;
  readonly DeployPipeline: CdkPipeline;
  readonly DeployProject: Project;

  constructor(galaxy: GalaxyExtension, props?: StackProps) {
    super(galaxy, 'CiCd', props);

    this.node.tryRemoveChild(this.Portal.node.id);
    this.Portal = new ImportedCiCdSolarSystem(this, this.Galaxy.Portal);

    this.DeployPipeline = new CdkPipeline(this, 'CdkPipeline', {
      name: `App-${this.Galaxy.Cosmos.Name}-Cdk-Pipeline`,
      cdkRepo: this.Galaxy.Cosmos.CdkRepo,
      deployRole: Role.fromRoleArn(this, 'CdkMasterRole', this.Galaxy.Cosmos.Portal.CdkMasterRoleStaticArn, {
        mutable: false,
      }),
      deployStacks: [this.RESOLVE(PATTERN.COSMOS, '*')],
    });
    this.DeployProject = this.DeployPipeline.Deploy;
  }
}
