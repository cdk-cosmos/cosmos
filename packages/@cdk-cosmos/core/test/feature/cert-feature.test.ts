import '@aws-cdk/assert/jest';
import * as path from 'path';
import { synthesizeStacks } from '../../../../../src/test';
import { App } from '@aws-cdk/core';
import { CosmosCoreStack, CosmosExtensionStack, GalaxyCoreStack } from '../../src';

const env1 = { account: 'account1', region: 'region' };
const env2 = { account: 'account2', region: 'region' };

const app = new App();
const cosmos = new CosmosCoreStack(app, 'Cos', {
  tld: 'cos.com',
  env: env1,
});
cosmos.addCert({ certsDir: path.join(__dirname, './certs') });

new GalaxyCoreStack(cosmos, 'Gal1', { cidr: '10.0.1.0/22' });
new GalaxyCoreStack(cosmos, 'Gal2', { cidr: '10.0.2.0/22', env: env2 });

const cosmosExtension = new CosmosExtensionStack(app, 'Test');
cosmosExtension.portal.addCert();
const [cosmosStack, certStack, cosmosExtensionStack] = synthesizeStacks(cosmos, cosmos.cert, cosmosExtension);

describe('Cert Feature', () => {
  test('should match snapshot', () => {
    expect(cosmosStack).toMatchSnapshot({
      Outputs: {
        CoreLibVersion: {
          Value: expect.any(String),
        },
      },
    });
    expect(certStack).toMatchSnapshot();
  });
});

describe('Cert Feature Extension', () => {
  test('should match snapshot', () => {
    expect(cosmosExtensionStack).toMatchSnapshot({
      Outputs: {
        AppLibVersion: {
          Value: expect.any(String),
        },
      },
    });
  });
});
