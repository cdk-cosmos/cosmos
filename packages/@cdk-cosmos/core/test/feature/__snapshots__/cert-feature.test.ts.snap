// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cert Feature Extension should match snapshot 1`] = `
Object {
  "Description": "Cosmos Extension: Singleton resources for the Cosmos, like CdkRepo and EcrRepo",
  "Outputs": Object {
    "AppLibVersion": Object {
      "Export": Object {
        "Name": "AppTestLibVersion",
      },
      "Value": "0.8.10",
    },
  },
}
`;

exports[`Cert Feature should match snapshot 1`] = `
Object {
  "Description": "Cosmos: Singleton resources for the Cosmos, like RootZone, CdkRepo and CdkMasterRole",
  "Outputs": Object {
    "CoreLibVersion": Object {
      "Export": Object {
        "Name": "CoreLibVersion",
      },
      "Value": "0.8.10",
    },
    "CrossAccountExportServiceToken": Object {
      "Export": Object {
        "Name": "CoreCrossAccountExportServiceToken",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "CustomCrossAccountExportsCustomResourceProviderHandler",
          "Arn",
        ],
      },
    },
    "RootZoneZoneId": Object {
      "Export": Object {
        "Name": "CoreRootZoneId",
      },
      "Value": Object {
        "Ref": "RootZone",
      },
    },
    "RootZoneZoneName": Object {
      "Export": Object {
        "Name": "CoreRootZoneName",
      },
      "Value": "cos.com",
    },
    "RootZoneZoneNameServers": Object {
      "Export": Object {
        "Name": "CoreRootZoneNameServers",
      },
      "Value": Object {
        "Fn::Join": Array [
          ",",
          Object {
            "Fn::GetAtt": Array [
              "RootZone",
              "NameServers",
            ],
          },
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6ArtifactHash": Object {
      "Description": "Artifact hash for asset \\"1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6\\"",
      "Type": "String",
    },
    "AssetParameters1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6S3Bucket": Object {
      "Description": "S3 bucket for asset \\"1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6\\"",
      "Type": "String",
    },
    "AssetParameters1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6S3VersionKey": Object {
      "Description": "S3 key for asset version \\"1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6\\"",
      "Type": "String",
    },
    "AssetParameterssourceHashArtifactHash": Object {
      "Description": "Artifact hash for asset \\"sourceHash\\"",
      "Type": "String",
    },
    "AssetParameterssourceHashS3Bucket": Object {
      "Description": "S3 bucket for asset \\"sourceHash\\"",
      "Type": "String",
    },
    "AssetParameterssourceHashS3VersionKey": Object {
      "Description": "S3 key for asset version \\"sourceHash\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CdkMasterRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "codebuild.amazonaws.com",
                  "codepipeline.amazonaws.com",
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
        "RoleName": "CoreCdkMasterRole",
        "Tags": Array [
          Object {
            "Key": "cosmos",
            "Value": "Cos",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Cert": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cosmos",
            "Value": "Cos",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.region.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParameters1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6S3Bucket",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6S3VersionKey",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters1fee2d2e3810eb71af5210fa50f48381a9c1eca9b4422d3c7e99870bb8217fd6S3VersionKey",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomCrossAccountExportsCustomResourceProviderHandler": Object {
      "DependsOn": Array [
        "CdkMasterRole",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterssourceHashS3Bucket",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterssourceHashS3VersionKey",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterssourceHashS3VersionKey",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CdkMasterRole",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "RootZone": Object {
      "Properties": Object {
        "HostedZoneConfig": Object {
          "Comment": "Core TLD Root Zone for Cos Cosmos.",
        },
        "HostedZoneTags": Array [
          Object {
            "Key": "cosmos",
            "Value": "Cos",
          },
        ],
        "Name": "cos.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
  },
}
`;

exports[`Cert Feature should match snapshot 2`] = `
Object {
  "Description": "Add Cert Bucket Features to the Cosmos",
  "Outputs": Object {
    "CertBucketBucketName": Object {
      "Export": Object {
        "Name": "CoreCertBucketName",
      },
      "Value": Object {
        "Ref": "CertBucket",
      },
    },
    "CombinedCertCrt": Object {
      "Export": Object {
        "Name": "CoreCombinedCertCrt",
      },
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://s3.region.",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "CertBucket",
            },
            "/combined-certs.crt",
          ],
        ],
      },
    },
    "CombinedCertPem": Object {
      "Export": Object {
        "Name": "CoreCombinedCertPem",
      },
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://s3.region.",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "CertBucket",
            },
            "/combined-certs.pem",
          ],
        ],
      },
    },
  },
  "Resources": Object {
    "CertBucket": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "core-cert-bucket",
        "Tags": Array [
          Object {
            "Key": "cosmos",
            "Value": "Cos",
          },
          Object {
            "Key": "cosmos:feature",
            "Value": "Cert",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "CertBucketPolicy": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "CertBucket",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::account1:root",
                    ],
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "CertBucket",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "CertBucket",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::account2:root",
                    ],
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "CertBucket",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "CertBucket",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C": Object {
      "DependsOn": Array [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "s3BucketName",
          "S3Key": "s3ObjectKey",
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "DeployFilesAwsCliLayer",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole",
            "Arn",
          ],
        },
        "Runtime": "python3.6",
        "Tags": Array [
          Object {
            "Key": "cosmos",
            "Value": "Cos",
          },
          Object {
            "Key": "cosmos:feature",
            "Value": "Cert",
          },
        ],
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "cosmos",
            "Value": "Cos",
          },
          Object {
            "Key": "cosmos:feature",
            "Value": "Cert",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "CertBucket",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "CertBucket",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy",
        "Roles": Array [
          Object {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DeployFilesAwsCliLayer": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": "s3BucketName",
          "S3Key": "s3ObjectKey",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "DeployFilesCustomResource": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucketName": Object {
          "Ref": "CertBucket",
        },
        "Prune": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C",
            "Arn",
          ],
        },
        "SourceBucketNames": Array [
          "s3BucketName",
        ],
        "SourceObjectKeys": Array [
          "s3ObjectKey",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
